connector/jobrunner/runner.py

```diff
@@ -0,0 +1,299 @@
+        self.has_connector = self._has_connector()
+        if self.has_connector:
+            self.has_channel = self._has_queue_job_column('channel')
+            self._initialize()
+
+    def close(self):
+        try:
+            self.conn.close()
+        except:
+            pass
+        self.conn = None
+
+    def _has_connector(self):
+        with closing(self.conn.cursor()) as cr:
+            try:
+                cr.execute("SELECT 1 FROM ir_module_module "
+                           "WHERE name=%s AND state=%s",
+                           ('connector', 'installed'))
+            except psycopg2.ProgrammingError as err:
+                if unicode(err).startswith('relation "ir_module_module" '
+                                           'does not exist'):
+                    return False
+                else:
+                    raise
+            return cr.fetchone()
+
+    def _has_queue_job_column(self, column):
+        if not self.has_connector:
+            return False
+        with closing(self.conn.cursor()) as cr:
+            cr.execute("SELECT 1 FROM information_schema.columns "
+                       "WHERE table_name=%s AND column_name=%s",
+                       ('queue_job', column))
+            return cr.fetchone()
+
+    def _initialize(self):
+        with closing(self.conn.cursor()) as cr:
+            # this is the trigger that sends notifications when jobs change
+            cr.execute("""
+                DROP TRIGGER IF EXISTS queue_job_notify ON queue_job;
+
+                CREATE OR REPLACE
+                    FUNCTION queue_job_notify() RETURNS trigger AS $$
+                BEGIN
+                    IF TG_OP = 'DELETE' THEN
+                        IF OLD.state != 'done' THEN
+                            PERFORM pg_notify('connector', OLD.uuid);
+                        END IF;
+                    ELSE
+                        PERFORM pg_notify('connector', NEW.uuid);
+                    END IF;
+                    RETURN NULL;
+                END;
+                $$ LANGUAGE plpgsql;
+
+                CREATE TRIGGER queue_job_notify
+                    AFTER INSERT OR UPDATE OR DELETE
+                    ON queue_job
+                    FOR EACH ROW EXECUTE PROCEDURE queue_job_notify();
+            """)
+            cr.execute("LISTEN connector")
+
+    def select_jobs(self, where, args):
+        query = "SELECT %s, uuid, id as seq, date_created, priority, eta, state " \
+                "FROM queue_job WHERE %s" % \
+                ('channel' if self.has_channel else 'NULL',
+                 where)
+        with closing(self.conn.cursor()) as cr:
+            cr.execute(query, args)
+            return list(cr.fetchall())
+
+    def set_job_enqueued(self, uuid):
+        with closing(self.conn.cursor()) as cr:
+            cr.execute("UPDATE queue_job SET state=%s, "
+                       "date_enqueued=date_trunc('seconds', now()::timestamp) "
```

Comment:
@sbidoul I just noticed that the now() should be at UTC. 
```sql
date_enqueued = date_trunc('seconds', now()::timestamp at time zone 'utc')
```
The code is therefore sensitive to the timezone defined on the database. To work as expected the timezone on the database should be set to 'UTC'
```sql
#show timezone;
 TimeZone 
----------
 UTC
(1 row)
```